README
Module: Readable Dates and Concealed Times Module
Version: 1.2.1
Author: Maurice Makaay and Oliver Riesen-Mallmann

Make dates easier to read for humans, by using relative indications like
"two hours ago", "yesterday, 12:30", "today, 17:52", "two weeks ago",
"last year", etc.

Conceals message timestamp for messages older than 24 hours. Don't let your
boss know that you wrote at work ...

All date indications can be customized through the language files.

Tested on version 5.2.20


License:
-------

Free to use, modify, or redistribute this code so long as I am given credit for
the original development somewhere. This code comes with absolutly no warranty.


Installation:
------------

1. Unpack the archive.

2. Move the directory "readable_dates" to the directory "mods" within your
   Phorum installation. Don't change the directory name!

3. Login as administrator in Phorum's administrative interface and go to the
   "Modules" section.

4. Enable the module "Readable Dates and Concealed Times".

5. Go to the settings screen for the module to configure the module (e.g. on
   what pages to replace the standard dates with readable dates).


Module settings:
---------------

== What granularity would you like to use?

By changing the granularity, you can set what interval you want to use as the
smallest one for formatting the dates. E.g. if you choose for seconds here,
then you might get "50 seconds ago" as the date. If you set granularity to
minutes, then the same date would show up as "this minute".

The interval that you select here will not be used for all formatted dates. So
you do not have to worry about getting things like "3024000 seconds ago" as a
date if you select seconds for the interval. For this example, it would be
"five weeks ago" instead.

The default value is "minutes".


== What is the time limit that you want to use for formatting?

If you set a time limit, then dates that are older than this time limit will
not be formatted by this module. For those, the default formatting that is
applied by Phorum will be used.

The default value is "no limit, format all dates".


== Enable readable numbers? (one, two, three instead of 1, 2, 3)

If you enable this feature, then numbers will be translated to their textual
representation. So instead of "3 days ago", "three days ago" will be shown.

The translation strings for the numbers can be found in the language files for
this module (mods/readable_dates/lang/*).

The default value is checked.


==== Select the pages for which you want to format the dates

== Format dates on the index page (list of forums)

The default value is checked.


== Format dates on the list page (list of messages)

The default value is checked.


== Format dates on the read page

The default value is unchecked.


== Format registration date (Registered: ... ago) for users on the read page

The default value is unchecked.


== Format dates on the user profile page

The default value is checked.


== Format dates on the private messages list page (list of messages)

The default value is checked.


== Format dates on the private messages read page

The default value is unchecked.


== Format dates on the buddy list page

The default value is checked.


== Format dates for the announcements module

The default value is checked.

==== Concealing Times Settings

== Conceal Times after 24 hours

The default value is unchecked.


Language support:
----------------

This module supports multiple languages.

The following languages are included:

- Dutch
- Durch informal
- English
- French
- German (Sie)
- German (Du) male
- German (Du) female
- Greek
- Russian
- Spanish

If you want to translate the module to a new language, you'll have to do the
following:

Copy lang/english.php to lang/yourlang.php (yourlang.php must have the same
name as Phorum's main language file that you use). After this, translate the
strings in lang/yourlang.php.

If you have created a translation for this module, please post a zip file
containing lang/yourlang.php in the module thread at phorum.org, so the
translation can be added to the module package. See:
http://www.phorum.org/phorum5/read.php?62,128691


Sanity Checks:
-------------

If you have problems with this module login as administrator in Phorum's
administrative interface and go to the "System Sanity Checks" section. Check
for warnings or errors for the Readable Dates and Concealed Times Module
Module.


Dependencies:
------------

- Do not use this module together with the "Conceal Message Timestamp Module".


Integration with other modules:
------------------------------

Other modules or scripts that want to make use of the readable date formatting,
can make use of a special hook call. If currently code like this is used for
formatting:

    $date = phorum_date($PHORUM["short_date_time"], $time);

then it can be replaced with this:

    if (!empty($PHORUM['hooks']['format_readable_date'])) {
        $date = phorum_hook('format_readable_date', $time);
    } else {
        $date = phorum_date($PHORUM['short_date_time'], $time);
    }


Contribution:
------------

Contribute on GitHub:

till version 1.1.11
https://github.com/mmakaay/Modules/tree/master/readable_dates

from version 1.1.12
https://github.com/langzeittest/readable_dates


Changelog
Changelog:
----------

2016-08-31 version 1.2.1 by Oliver Riesen-Mallmann <oliver@riesen.org>
- Made XHTML 1.0 Transitional compliant.

2016-07-27 version 1.2.0 by Oliver Riesen-Mallmann <oliver@riesen.org>
- Module renamed to "Readable Dates and Concealed Times Module".
- Functions of the "Conceal Message Timestamp Module" included. Formatting
  date and times in two separate modules didn't work well.
- Fixed formatting for the announcement module.
- README extended (license, module settings, language support, dependencies
  and GitHub link added).
- Performance improvements.
- Sanity checks added.

2015-09-18 version 1.1.12 by Oliver Riesen <oliver@riesen.org>
- Fixed undefined index warning.
- Performance improvements.
- Added Greek language file. Thanks to Thanos.

2011-01-09 version 1.1.11 by Maurice Makaay
- Added new German language files. Thanks to Oliver Riesen!

2010-04-26 version 1.1.10
- The original dates are now available as $data['orig_<datefield>'], so they
  can still be accessed when needed from the templates.

2009-08-26 version 1.1.9
- Added support for the Russian language. Also added a Russian language file.
  Thanks to phorum.org user sailord for the translation!
- Added a Spanish translation. Thanks to Chris Brown!
- Added a French translation. Thanks to phorum.org user manuvb!

2008-12-14 version 1.1.8
- Fixed a typo that caused "date_last_active" for the user profile screen to
  never be formatted by this module.

2008-07-25 version 1.1.7
- Added a switch in the module config to enable/disable the formatting of the
  user registration date on the read pages (Registered: ... ago). If this
  switch is enabled, then dates after the configured time limit will be
  formatted using a standard short date. This is different from Phorum's
  default behavior, because by default Phorum would use a readable date-like
  formatting for this date already.

2008-05-28 version 1.1.6
- Another undefined index warning fixed and again it was CBiLL who was kind
  enough to be the lab rat.

2008-03-26 version 1.1.5
- Some more undefined index warnings fixed. These would occur if user activity
  tracking is disabled or if the user hides its activity through the privacy
  settings. Thanks to CBiLL for the report!

- A configuration option was added to make it possible to disable the readable
  numbers (one, two, three instead of 1, 2, 3).

2008-03-11 version 1.1.4
- Fixes for some undefined index warnings in readable_dates.php. Thanks to
  CBiLL for the heads up!

2008-03-03 version 1.1.3
- Added support for formatting dates in the private message and buddy
  interfaces (pm list, pm read, buddy list). These features need Phorum version
  5.2.7 or higher (the admin interface will warn about this if an older version
  of Phorum is used).

2008-03-02 version 1.1.2
- Added an option to specify the maximum time for which a date should be
  formatted by this module. Dates that are older will be displayed using their
  original formatting. Thanks to CBiLL for the feature suggestion.
- Fixed a bug, which resulted in empty dates being formatted as if the date was
  Januari 1st, 1970 (Epoch zero). Thanks to CBiLL for the bug report.

2008-03-01 version 1.1.1
- Added a better German translation to the package. Thanks to Thomas for fixing
  my crappy German translation!
- Fixed the formatting code to make sure that it will work correctly in
  localized environments.

2008-03-01 version 1.1.0
- Rewrote the module to be compatible with 5.2 and to make full use of new
  Phorum features.

2006-05-08 version 1.0.7
- Fixed typo: "Las edit was" -> "Last edit was" (thanks CBiLL).

2006-03-28 version 1.0.6
- Do a restore for time formats at the end of the list and the read hook, as
  suggested by Oliver Riesen. This will take away some problems with other
  modules that are using time formatting as well (but probably not all of them,
  in case date formatting is run before the list/read hooks).

2006-03-01 version 1.0.5
- The script did not load the right defaults.php file.
- Even more fixes for TZ handling. Users that have a timezone set in the
  control center should now get a correct date too when using this mod.

2006-01-20 version 1.0.4
- Some more fixes for correct TZ handling. The previous version was trying to
  fix problems in the wrong place, which caused unexpected side-effect
  problems.
- The dates on the read page can now be formatted as well.
- A settings screen has been added. In the settings you can configure on which
  pages you want to apply the readable date formatting.

2006-01-19 version 1.0.3
- Some fixes for correct TZ handling (as suggested by regdos)

2006-01-17 version 1.0.2
- wrong datafield used for lastpost formatting (reported by regdos)

2006-01-16 version 1.0.1
- Minor fix in French and German language packages

2006-01-10 version 1.0.0
- Initial release



settings.php
<?php

if (!defined('PHORUM_ADMIN')) return;

require_once './mods/readable_dates/defaults.php';

// Save settings
if (count($_POST)) {
    foreach (array(
        'buddies', 'conceal_times', 'index', 'list', 'mod_announcements', 'pm_list',
        'pm_read', 'profile', 'read', 'read_userregdate', 'readable_numbers'
    ) as $page) {
        $PHORUM['mod_readable_dates'][$page] = empty($_POST[$page]) ? 0 : 1;
    }
    $PHORUM['mod_readable_dates']['granularity'] = (int)$_POST['granularity'];
    $PHORUM['mod_readable_dates']['max_time'] = empty($_POST['max_time']) ? '' : (int)$_POST['max_time'];

    phorum_db_update_settings(array(
        'mod_readable_dates' => $PHORUM['mod_readable_dates']
    ));
    phorum_admin_okmsg('Settings updated');
}

include_once './include/admin/PhorumInputForm.php';
$frm = new PhorumInputForm ('', 'post', 'Save settings');
$frm->hidden('module', 'modsettings');
$frm->hidden('mod', 'readable_dates');

$frm->addbreak('Edit settings for the Readable Dates and Concealed Times module');

$frm->addbreak('Readable Dates Settings');

$row = $frm->addrow(
    'What granularity would you like to use?',
    $frm->select_tag(
        'granularity',
        array(
            0        => 'seconds',
            60       => 'minutes',
            3600     => 'hours',
            86400    => 'days',
            604800   => 'weeks',
            2592000  => 'months',
            30758400 => 'years'
        ),
        $PHORUM['mod_readable_dates']['granularity']
    )
);
$frm->addhelp($row, 'What granularity would you like to use?',
    'By changing the granularity, you can set what interval you want
     to use as the smallest one for formatting the dates. E.g. if you
     choose for seconds here, then you might get &quot;50 seconds ago&quot;
     as the date. If you set granularity to minutes, then the same date
     would show up as &quot;this minute&quot;.<br /><br />
     The interval that you select here will not be used for all formatted
     dates. So you do not have to worry about getting things like
     &quot;3024000 seconds ago&quot; as a date if you select seconds for the
     interval. For this example, it would be &quot;five weeks ago&quot; instead.'
);

$row = $frm->addrow(
    'What is the time limit that you want to use for formatting?',
    $frm->select_tag(
        'max_time',
        array(
            0          => 'no limit, format all dates',
            3600       => ' 1 hour',
            7200       => ' 2 hours',
            10800      => ' 3 hours',
            14400      => ' 4 hours',
            28800      => ' 8 hours',
            43200      => '12 hours',
            86400      => ' 1 day',
            172800     => ' 2 days',
            259200     => ' 3 days',
            345600     => ' 4 days',
            432000     => ' 5 days',
            518400     => ' 6 days',
            604800     => ' 1 week',
            1209600    => ' 2 weeks',
            1814400    => ' 3 weeks',
            2419200    => ' 4 weeks',
            2628000    => ' 1 month',
            5256000    => ' 2 months',
            7884000    => ' 3 months',
            10512000   => ' 4 months',
            13140000   => ' 5 months',
            15724800   => ' 6 months',
            18396000   => ' 7 months',
            21024000   => ' 8 months',
            23652000   => ' 9 months',
            26280000   => '10 months',
            28908000   => '11 months',
            31536000   => ' 1 year'
        ),
        $PHORUM['mod_readable_dates']['max_time']
    )
);
$frm->addhelp($row, 'Time limit to use for formatting',
    'If you set a time limit, then dates that are older than
     this time limit will not be formatted by this module.
     For those, the default formatting that is applied by Phorum
     will be used.'
);

$row = $frm->addrow(
    'Enable readable numbers? (one, two, three instead of 1, 2, 3)',
    $frm->checkbox(
        'readable_numbers', '1', 'Yes',
        $PHORUM['mod_readable_dates']['readable_numbers']
    )
);
$frm->addhelp($row, 'Enable readable numbers?',
    'If you enable this feature, then numbers will be translated to their
     textual representation. So instead of &quot;3 days ago&quot;, &quot;three days
     ago&quot; will be shown.<br /><br />
     The translation strings for the numbers can be found in the
     language files for this module (mods/readable_dates/lang/*).'
);

$frm->addbreak('Select the pages for which you want to format the dates');

$frm->addrow(
    'Format dates on the index page (list of forums)',
    $frm->checkbox(
        'index', '1', 'Yes',
        $PHORUM['mod_readable_dates']['index']
    )
);
$frm->addrow(
    'Format dates on the list page (list of messages)',
    $frm->checkbox(
        'list', '1', 'Yes',
        $PHORUM['mod_readable_dates']['list']
    )
);
$frm->addrow(
    'Format dates on the read page',
    $frm->checkbox(
        'read', '1', 'Yes',
        $PHORUM['mod_readable_dates']['read']
    )
);
$frm->addrow(
    'Format registration date (Registered: ... ago) for users on the read page',
    $frm->checkbox(
        'read_userregdate', '1', 'Yes',
        $PHORUM['mod_readable_dates']['read_userregdate']
    )
);
$frm->addrow(
    'Format dates on the user profile page',
    $frm->checkbox(
        'profile', '1', 'Yes',
        $PHORUM['mod_readable_dates']['profile']
    )
);

$frm->addrow(
    'Format dates on the private messages list page (list of messages)',
    $frm->checkbox(
        'pm_list', '1', 'Yes',
        $PHORUM['mod_readable_dates']['pm_list']
    )
);
$frm->addrow(
    'Format dates on the private messages read page',
    $frm->checkbox(
        'pm_read', '1', 'Yes',
        $PHORUM['mod_readable_dates']['pm_read']
    )
);
$frm->addrow(
    'Format dates on the buddy list page',
    $frm->checkbox(
        'buddies', '1', 'Yes',
        $PHORUM['mod_readable_dates']['buddies']
    )
);

// This one is special. The announcement module should handle this
// by calling the readable dates formatting code, but since this is
// a core module, we'll make an exception and handle the data
// formatting directly from this module.
$frm->addrow(
    'Format dates for the announcements module',
    $frm->checkbox(
        'mod_announcements', '1', 'Yes',
        $PHORUM['mod_readable_dates']['mod_announcements']
    )
);

$frm->addbreak('Concealing Times Settings');

$frm->addrow(
    'Conceal Times after 24 hours',
    $frm->checkbox(
        'conceal_times', '1', 'Yes',
        $PHORUM['mod_readable_dates']['conceal_times']
    )
);

$frm->show();
?>
